File: lib/sidereon/gnss/broadcast_comparison.ex
defmodule Sidereon.GNSS.BroadcastComparison do
  @moduledoc """
  Broadcast-ephemeris accuracy: compare a broadcast navigation product against a
  precise SP3 product over a window.

  This is the standard broadcast-orbit / clock accuracy check (the orbit and clock
  pieces of the signal-in-space range error, SISRE). For each satellite at each
  epoch it differences the broadcast-evaluated ECEF position and clock against the
  precise SP3 values, decomposes the position error into radial / along-track /
  cross-track (RAC) components, and summarizes the differences as RMS and maximum
  statistics per satellite and overall.

  Both products are evaluated through `Sidereon.GNSS.Ephemeris`, so the frame
  (ITRF/IGS ECEF, meters), the time scale (GPST), and the clock sign convention
  (positive = satellite clock ahead of system time) are exactly as documented
  there. Only epochs where **both** sources return a valid state contribute to the
  statistics; an epoch missing from either product is skipped, never extrapolated.

  ## RAC frame

  The radial/along-track/cross-track unit vectors are built from the **precise**
  state: radial along the position vector, cross-track along the orbital angular
  momentum `r x v`, and along-track completing the right-handed triad. The SP3
  position/clock interpolation does not expose a velocity, so the velocity is
  derived by a centered finite difference of the precise position
  (`(r(t+dt) - r(t-dt)) / 2dt`, falling back to a one-sided difference at a window
  edge). The position-difference vector `broadcast - precise` is projected onto
  this triad.

  ## Expected magnitudes

  GPS LNAV broadcast orbits differ from IGS precise orbits at roughly the 1-2 m
  RMS level (3D), dominated by the along-track and radial components; Galileo and
  BeiDou MEO are comparable. A result far outside this band (tens of meters or
  more) indicates a parse or evaluation defect rather than normal broadcast error.

  The broadcast models follow IS-GPS-200 (GPS LNAV), the Galileo OS-SIS-ICD, and
  the BeiDou BDS-SIS-ICD; the precise product is SP3-c / SP3-d (IGS).

  ## Example

      {:ok, broadcast} = Sidereon.GNSS.Broadcast.load("BRDC.rnx")
      {:ok, sp3} = Sidereon.GNSS.SP3.load("igs.sp3")

      report =
        Sidereon.GNSS.BroadcastComparison.compare(broadcast, sp3, ["G01", "E11"], %{
          from: ~N[2020-06-25 02:00:00],
          to: ~N[2020-06-25 04:00:00],
          step_s: 300
        })

      report.overall.orbit_3d_rms_m   # 3D orbit RMS over all satellites, meters
      report.per_satellite["G01"].radial_rms_m

  ## Broadcast and precise products on a sample day

  Comparing the GPS LNAV broadcast message against the GBM precise SP3 product for
  2020 day-of-year 177, all GPS satellites over a multi-hour window at a 15 min
  step, gives an overall 3D orbit RMS of about **1.5 m** (max ~4 m), split as
  roughly 1.1 m radial / 0.9 m along-track / 0.5 m cross-track, the expected GPS
  broadcast accuracy. The raw clock differences (`clock_rms_m`) are larger
  (several meters) because the broadcast and precise clocks are referenced to
  different time datums, which differ by a common per-epoch offset that drifts
  over the day. `clock_datum_removed_rms_m` removes that common offset (the
  per-epoch median over satellites) and reports the actual signal-in-space clock
  error, which is several times smaller.
  The same call works against a broadcast product with no change of shape, so
  `mix gnss.broadcast_diff --nav BRDC.rnx --sp3 igs.sp3 --from ... --to ...`
  prints this table from the command line.

  The orbit/clock differencing, RAC decomposition, finite-difference velocity, and
  the RMS/median/datum statistics are computed in the Rust core; this module
  validates inputs, marshals the per-epoch evaluation keys across the NIF, and
  rebuilds the public report.
  """

  alias Sidereon.GNSS.Broadcast
  alias Sidereon.GNSS.Ephemeris
  alias Sidereon.GNSS.SP3
  alias Sidereon.GNSS.Time
  alias Sidereon.NIF

  defmodule Stats do
    @moduledoc """
    Orbit and clock difference statistics for one satellite (or the overall set).

    All values are meters except `count` (the number of compared epochs).
    `orbit_3d_rms_m` / `orbit_3d_max_m` are the Euclidean position-difference
    magnitudes. `radial_*`, `along_*`, `cross_*` are the RMS and max of the signed
    RAC components of the position difference (`broadcast - precise`).
    `clock_rms_m` / `clock_max_m` are the **raw** satellite-clock differences
    scaled to meters by the speed of light; they are `nil` when neither product
    carried a clock estimate for any compared epoch. They are dominated by the
    per-epoch common reference-clock offset between the two products' time
    datums. `clock_datum_removed_rms_m` / `clock_datum_removed_max_m` are the
    same differences after that per-epoch common offset (the median over all
    satellites at the epoch) is removed, the actual signal-in-space clock error
    (the SISRE clock term), typically several times smaller than the raw value.
    """
    @enforce_keys [
      :count,
      :orbit_3d_rms_m,
      :orbit_3d_max_m,
      :radial_rms_m,
      :radial_max_m,
      :along_rms_m,
      :along_max_m,
      :cross_rms_m,
      :cross_max_m,
      :clock_rms_m,
      :clock_max_m,
      :clock_datum_removed_rms_m,
      :clock_datum_removed_max_m
    ]
    defstruct [
      :count,
      :orbit_3d_rms_m,
      :orbit_3d_max_m,
      :radial_rms_m,
      :radial_max_m,
      :along_rms_m,
      :along_max_m,
      :cross_rms_m,
      :cross_max_m,
      :clock_rms_m,
      :clock_max_m,
      :clock_datum_removed_rms_m,
      :clock_datum_removed_max_m
    ]

    @type t :: %__MODULE__{
            count: non_neg_integer(),
            orbit_3d_rms_m: float() | nil,
            orbit_3d_max_m: float() | nil,
            radial_rms_m: float() | nil,
            radial_max_m: float() | nil,
            along_rms_m: float() | nil,
            along_max_m: float() | nil,
            cross_rms_m: float() | nil,
            cross_max_m: float() | nil,
            clock_rms_m: float() | nil,
            clock_max_m: float() | nil,
            clock_datum_removed_rms_m: float() | nil,
            clock_datum_removed_max_m: float() | nil
          }
  end

  defmodule Report do
    @moduledoc """
    The result of a broadcast and precise product comparison.

    `per_satellite` maps each satellite id to its `Stats`; `overall` aggregates
    every compared epoch across all satellites. `missing` lists `{satellite_id,
    count}` pairs counting epochs that were skipped because one or both products
    had no valid state there.
    """
    @enforce_keys [:per_satellite, :overall, :missing]
    defstruct [:per_satellite, :overall, :missing]

    @type t :: %__MODULE__{
            per_satellite: %{String.t() => Stats.t()},
            overall: Stats.t(),
            missing: [{String.t(), non_neg_integer()}]
          }
  end

  @doc """
  Compare a broadcast product against a precise SP3 product over `window`.

  `broadcast` is an `Sidereon.GNSS.Broadcast` handle and `precise` an
  `Sidereon.GNSS.SP3` handle for the same day; `sat_ids` is a list of canonical
  RINEX tokens; `window` is the `Sidereon.GNSS.Ephemeris` window map (`:from`,
  `:to`, `:step_s`). Returns an `Sidereon.GNSS.BroadcastComparison.Report`.

  The window should sit within both the SP3 file span and the broadcast records'
  fit intervals; epochs missing from either product are counted in
  `report.missing` and excluded from the statistics.
  """
  @spec compare(Broadcast.t(), SP3.t(), [String.t()], Ephemeris.window()) :: Report.t()
  def compare(%Broadcast{handle: broadcast}, %SP3{handle: precise}, sat_ids, window)
      when is_list(sat_ids) do
    epochs = window_epochs(window)
    half = round(window.step_s / 2.0)
    epoch_keys = Enum.map(epochs, &epoch_keys(&1, half))

    {overall, per_satellite, missing} =
      NIF.broadcast_comparison(broadcast, precise, sat_ids, epoch_keys, half / 1.0)

    %Report{
      overall: decode_stats(overall),
      per_satellite: Map.new(per_satellite, fn {sat, stats} -> {sat, decode_stats(stats)} end),
      missing: Enum.map(missing, fn {sat, count} -> {sat, count} end)
    }
  end

  # The per-epoch evaluation keys the core consumes: the broadcast query is a
  # continuous J2000 second (the same value `Sidereon.GNSS.Broadcast.position`
  # marshals), the precise queries are split Julian dates for the epoch and the
  # +/-half-step velocity finite-difference neighbours.
  defp epoch_keys(epoch, half) do
    {jd_whole, jd_fraction} = Time.epoch_to_split_jd(epoch)
    {jd_whole_p, jd_fraction_p} = Time.epoch_to_split_jd(NaiveDateTime.add(epoch, half, :second))
    {jd_whole_m, jd_fraction_m} = Time.epoch_to_split_jd(NaiveDateTime.add(epoch, -half, :second))
    {:ok, broadcast_j2000_s} = Time.epoch_to_j2000_seconds_fractional(epoch)

    {broadcast_j2000_s, jd_whole, jd_fraction, jd_whole_p, jd_fraction_p, jd_whole_m,
     jd_fraction_m}
  end

  defp decode_stats(
         {count,
          [
            orbit_3d_rms_m,
            orbit_3d_max_m,
            radial_rms_m,
            radial_max_m,
            along_rms_m,
            along_max_m,
            cross_rms_m,
            cross_max_m,
            clock_rms_m,
            clock_max_m,
            clock_datum_removed_rms_m,
            clock_datum_removed_max_m
          ]}
       ) do
    %Stats{
      count: count,
      orbit_3d_rms_m: orbit_3d_rms_m,
      orbit_3d_max_m: orbit_3d_max_m,
      radial_rms_m: radial_rms_m,
      radial_max_m: radial_max_m,
      along_rms_m: along_rms_m,
      along_max_m: along_max_m,
      cross_rms_m: cross_rms_m,
      cross_max_m: cross_max_m,
      clock_rms_m: clock_rms_m,
      clock_max_m: clock_max_m,
      clock_datum_removed_rms_m: clock_datum_removed_rms_m,
      clock_datum_removed_max_m: clock_datum_removed_max_m
    }
  end

  defp window_epochs(%{from: from, to: to, step_s: step_s}) do
    total_s = NaiveDateTime.diff(to, from, :second)
    n = div(total_s, step_s)
    for i <- 0..n//1, do: NaiveDateTime.add(from, i * step_s, :second)
  end
end

File: native/sidereon_nif/src/broadcast_comparison.rs
//! Rustler boundary for broadcast-vs-precise ephemeris accuracy (SISRE).
//!
//! Pure glue over `sidereon_core::broadcast_comparison`: it decodes the
//! broadcast and precise SP3 resource handles plus the sampling window the
//! Sidereon interface marshals (the broadcast J2000-second span, the precise
//! split Julian date at the window start, the step, and the velocity half-step),
//! calls the crate's window-form comparison driver (which builds the per-epoch
//! grid internally), and encodes the per-satellite / overall / missing report
//! back. No grid construction, RAC projection, statistics, or datum removal live
//! here.

use crate::broadcast::BroadcastResource;
use crate::sp3::Sp3Resource;
use rustler::{Encoder, Env, Error, NifResult, ResourceArc, Term};
use sidereon_core::astro::time::model::JulianDateSplit;
use sidereon_core::broadcast_comparison::{compare_window, CompareStats, CompareWindow};
use sidereon_core::{GnssSatelliteId, GnssSystem};

/// Parse a canonical RINEX satellite token (e.g. `"G05"`) into a typed id.
fn parse_token(token: &str) -> NifResult<GnssSatelliteId> {
    let mut chars = token.chars();
    let letter = chars
        .next()
        .ok_or_else(|| Error::Term(Box::new("empty satellite token")))?;
    let system = GnssSystem::from_letter(letter)
        .ok_or_else(|| Error::Term(Box::new(format!("unknown GNSS system letter {letter:?}"))))?;
    let prn: u8 = chars
        .as_str()
        .parse()
        .map_err(|_| Error::Term(Box::new(format!("invalid satellite PRN in {token:?}"))))?;
    GnssSatelliteId::new(system, prn).map_err(crate::errors::invalid_input)
}

/// Encode one statistics record as `{count, [12 optional floats]}` where each
/// float is `nil` when absent, in the fixed order the Elixir wrapper decodes.
fn encode_stats<'a>(env: Env<'a>, stats: &CompareStats) -> Term<'a> {
    let opt = |value: Option<f64>| -> Term<'a> {
        match value {
            Some(v) => v.encode(env),
            None => rustler::types::atom::nil().encode(env),
        }
    };
    let fields = vec![
        opt(stats.orbit_3d_rms_m),
        opt(stats.orbit_3d_max_m),
        opt(stats.radial_rms_m),
        opt(stats.radial_max_m),
        opt(stats.along_rms_m),
        opt(stats.along_max_m),
        opt(stats.cross_rms_m),
        opt(stats.cross_max_m),
        opt(stats.clock_rms_m),
        opt(stats.clock_max_m),
        opt(stats.clock_datum_removed_rms_m),
        opt(stats.clock_datum_removed_max_m),
    ];
    (stats.count as u64, fields).encode(env)
}

/// Compare a broadcast product against a precise SP3 product over a sampling
/// window. The interface supplies the two start anchors (the broadcast J2000
/// second axis `(t0, t1)` and the precise split Julian date at `t0`), the step,
/// and the velocity half-step; the core driver builds the per-epoch grid.
/// Returns `{overall_stats, per_satellite, missing}` where `per_satellite` is a
/// list of `{sat_token, stats}` and `missing` a list of `{sat_token,
/// skipped_count}`. Dirty-CPU: a full IGS day across all satellites is unbounded
/// relative to the 1 ms NIF budget.
#[rustler::nif(schedule = "DirtyCpu")]
#[allow(clippy::too_many_arguments)]
fn broadcast_comparison<'a>(
    env: Env<'a>,
    broadcast: ResourceArc<BroadcastResource>,
    precise: ResourceArc<Sp3Resource>,
    satellites: Vec<String>,
    broadcast_t0_j2000_s: f64,
    broadcast_t1_j2000_s: f64,
    precise_start_jd_whole: f64,
    precise_start_fraction: f64,
    step_s: f64,
    velocity_half_s: f64,
) -> NifResult<Term<'a>> {
    let satellites: Vec<GnssSatelliteId> = satellites
        .iter()
        .map(|token| parse_token(token))
        .collect::<NifResult<_>>()?;

    let window = CompareWindow {
        broadcast_window_j2000_s: (broadcast_t0_j2000_s, broadcast_t1_j2000_s),
        precise_start: JulianDateSplit::new(precise_start_jd_whole, precise_start_fraction)
            .map_err(crate::errors::invalid_input)?,
        step_s,
        velocity_half_s,
    };

    let report = compare_window(&broadcast.store, &precise.sp3, &satellites, &window)
        .map_err(crate::errors::invalid_input)?;

    let per_satellite: Vec<(String, Term<'a>)> = report
        .per_satellite
        .iter()
        .map(|(sat, stats)| (sat.to_string(), encode_stats(env, stats)))
        .collect();
    let missing: Vec<(String, u64)> = report
        .missing
        .iter()
        .map(|(sat, count)| (sat.to_string(), *count as u64))
        .collect();

    Ok((encode_stats(env, &report.overall), per_satellite, missing).encode(env))
}
